src/plugins/pcr_primer_design_for_dna_assembly/src/PCRPrimerDesignForDNAAssemblyOPWidget.cpp

```diff
@@ -0,0 +1,236 @@
+
+    U2WidgetStateStorage::restoreWidgetState(savableWidget);
+
+    sl_updateMaxValues();
+    for (auto minSb : minMaxValuesSb.keys()) {
+        connect(minSb, &QSpinBox::editingFinished, this, &PCRPrimerDesignForDNAAssemblyOPWidget::sl_minValueChanged);
+        connect(minMaxValuesSb[minSb], &QSpinBox::editingFinished, this, &PCRPrimerDesignForDNAAssemblyOPWidget::sl_maxValueChanged);
+    }
+    connect(tbLoadBackbone, &QAbstractButton::clicked, this, &PCRPrimerDesignForDNAAssemblyOPWidget::sl_loadBackbone);
+    connect(tbsaveRandomSequences, &QAbstractButton::clicked, this, &PCRPrimerDesignForDNAAssemblyOPWidget::sl_saveRandomSequences);
+    connect(tbLoadOtherSequencesInPcr, &QAbstractButton::clicked, this, &PCRPrimerDesignForDNAAssemblyOPWidget::sl_loadOtherSequenceInPcr);
+}
+
+void PCRPrimerDesignForDNAAssemblyOPWidget::sl_start() {
+    //TODO
+}
+
+void PCRPrimerDesignForDNAAssemblyOPWidget::sl_selectManually() {
+    auto currentSMButton = qobject_cast<QToolButton*>(sender());
+    SAFE_POINT(currentSMButton != nullptr, L10N::nullPointerError("QAbstractButton"), );
+
+    if (smButton != nullptr) {
+        bool disableState = true;
+        if (smButton->isChecked()) {
+            smButton->setChecked(false);
+            disableState = false;
+        }
+        smButton = nullptr;
+        start = nullptr;
+        end = nullptr;
+        disconnect(updateRegionConnection);
+        updateRegionConnection = QMetaObject::Connection();
+
+        if (disableState) {
+            return;
+        }
+    }
+
+    smButton = currentSMButton;
+    if (smButton == tbLeftAreaSelectManually) {
+        start = sbLeftAreaStart;
+        end = sbLeftAreaEnd;
+    } else {
+        start = sbRightAreaStart;
+        end = sbRightAreaEnd;
+    }
+    auto activeSequenceContext = annDnaView->getActiveSequenceContext();
+    SAFE_POINT(activeSequenceContext != nullptr, L10N::nullPointerError("ADVSequenceObjectContext"), );
+
+    auto sequenceSelection = activeSequenceContext->getSequenceSelection();
+    SAFE_POINT(sequenceSelection != nullptr, L10N::nullPointerError("DNASequenceSelection"), );
+
+    updateRegionConnection = connect(sequenceSelection, &DNASequenceSelection::si_onSelectionChanged, this, &PCRPrimerDesignForDNAAssemblyOPWidget::sl_updateRegion);
+}
+
+void PCRPrimerDesignForDNAAssemblyOPWidget::sl_updateRegion() {
+    auto sequenceSelection = qobject_cast<DNASequenceSelection*>(sender());
+    SAFE_POINT(sequenceSelection != nullptr, L10N::nullPointerError("DNASequenceSelection"), );
+
+    const auto& selectedRegions = sequenceSelection->getSelectedRegions();
+    CHECK(selectedRegions.size() != 0, );
+
+    if (selectedRegions.size() > 1) {
+        coreLog.details(tr("Multiple regions selection, the only first is set as primer search area."));
+    }
+    const auto& region = selectedRegions.first();
+    start->setValue(region.startPos + 1);
```

Comment:
start & end are nullable. Here is a good place for a SAFE_POINT

Reply:
OK